cmd/compute-domain-kubelet-plugin/device_state.go

```diff
@@ -153,42 +156,57 @@ func (s *DeviceState) Prepare(ctx context.Context, claim *resourceapi.ResourceCl
 		return nil, fmt.Errorf("unable to create CDI spec file for claim: %w", err)
 	}
 
-	preparedClaims[claimUID] = preparedDevices
+	// Add ResourceClaim API object to checkpoint so that for Unprepare() only
+	// local state is required (ResourceClaim object might have been deleted
+	// from the API server).
+	preparedClaims[claimUID] = *claim
+	preparedDevicesByClaimUID[claimUID] = preparedDevices
 	if err := s.checkpointManager.CreateCheckpoint(DriverPluginCheckpointFile, checkpoint); err != nil {
-		return nil, fmt.Errorf("unable to sync to checkpoint: %v", err)
+		return nil, fmt.Errorf("unable to create checkpoint: %w", err)
 	}
+	klog.Infof("checkpoint written for claim %v", claimUID)
 
-	return preparedClaims[claimUID].GetDevices(), nil
+	return preparedDevicesByClaimUID[claimUID].GetDevices(), nil
 }
 
-func (s *DeviceState) Unprepare(ctx context.Context, claim *resourceapi.ResourceClaim) error {
+func (s *DeviceState) Unprepare(ctx context.Context, claimRef kubeletplugin.NamespacedObject) error {
 	s.Lock()
 	defer s.Unlock()
 
-	claimUID := string(claim.UID)
-
-	if err := s.unprepareDevices(ctx, claim); err != nil {
-		return fmt.Errorf("unprepare devices failed: %w", err)
-	}
+	claimUID := string(claimRef.UID)
 
+	// Rely on local checkpoint state for ability to clean up.
```

Comment:
On reliance on only checkpointing data: of course, checkpointing data is vulnerable to node failures (data may be lost in rare cases). But device state can be assumed to get lost / reset 'in sync' (e.g., as part of recovery from a failure affecting the checkpoint data it is likely for a reboot to happen).